se/se_epub_build.py

```diff
@@ -816,7 +816,8 @@ def build(self, metadata_xhtml: str, metadata_tree: se.easy_xml.EasyXmlTree, run
 				file.truncate()
 
 			# Rebuild the NCX
-			toc_tree = se.epub.convert_toc_to_ncx(work_epub_root_directory, toc_filename, navdoc2ncx_xsl_filename)
+			with importlib_resources.path("se.data", "navdoc2ncx.xsl") as navdoc2ncx_xsl_filename:
```

Comment:
I'm curious why you use a one-line `with` here and in some other places, instead of inlining the output of `importlib_resources.path()`?

In other words, why not:

````python
toc_tree = se.epub.convert_toc_to_ncx(work_epub_root_directory, toc_filename, importlib_resources.path("se.data", "navdoc2ncx.xsl"))
````

Reply:
According to the docs, the `path()` function returns a context manager which is used for cleaning up temporary files that are created if the resource needs to be extracted from an archive. Without the `with` statement you need to explicitly call a `close()` function or you will leak files.

The old API did the cleanup implicitly so a cleanup function didn't need to be called and a context manager wasn't necessary.

Reply:
That is an API that's so crazy I didn't even think to look it up. Expect a path, get a context manager for a path. Wild.

Is there a way we can get around this? Using a context manager to essentially assign a path to a variable seems... unpleasant. Dare I say it, unpythonic even.

Reply:
What about doing something like:

````python
path = Path(os.path.dirname(__file__)) / "data" / "templates" / "LICENSE.md"
````
If something like that works we could store `os.path.dirname(__file__)` into a const in `__init.py`. I did the research on package resource lookup a long time ago so it's a bit fuzzy for me why I went with `pkg_resources` instead of the above method.

Reply:
The tricky bit is that a resource in a package is not always a real file on the local filesystem. A package could be in an archive (.zip or .tar.bz or whatever) or even streamed over a network, depending on the context. The whole point of `pkg_resources` and `importlib.resources` is to abstract away some of that complexity so you don't have to deal with it.

That being said, the package resources are _usually_ just files on the local filesystem, so the idea of getting the dirname of `__file__` would probably work fine until one day it didn't and then you'd have to go back to one of the libraries.

Reply:
OK. In our case, we know that all of our data files are not zip files and it would be highly unlikely for them to be on a separate root fs/network share than `__init__.py`, and in the crazy case that they are we can just say it's not supported.

If there's no obvious gotchas to using `os.path.dirname(__file__)` then I think I prefer that as a more obvious and elegant approach, if not the 100% airtight approach. We could wrap it in a helper function in `__init_py`, maybe something like `get_data_file_path() -> Path` so that if one day it stops working we can replace that function instead of the rooting around the whole codebase.

Thoughts?

(IMHO the package resource API should have different methods for accessing the paths of plain files and zip files, much like there are different methods for accessing files for unicode/string reading and binary reading. But, at the end of the day that's not our problem.)

Reply:
Actually, now that I wrote that out, why not have a `get_data_file_path() -> Path` function that wraps the context manager from importlib and returns a `Path`? That way we get to use importlib, but without having to do a 2-line `with` incantation just to assign a variable.

Reply:
It's not really in your control whether the standardebooks package is zipped or not. The default python importer supports files in zip format if someone chooses to import your code that way.

For example a wheel is .zip format, so I can import and run your code without ever installing the wheel:

    $ python3 setup.py bdist_wheel
    $ PYTHONPATH=dist/standardebooks-1.0.28-py3-none-any.whl python3 -c "import se.executables; se.executables.se_help()"
    The following commands are available:
    british2american
    build
    build-images
    ...

Not sure why I would want to, but the point is Python supports it.

Python also supports custom loaders for imports so if someone is using a custom loader for some reason then all bets are off about whether your files are on the filesystem or not. The loader could be streaming the code directly from a package server.

So it really just comes down to how robust you want your packaging to be. Most of the time it will probably be fine to look for resources on the filesystem. In some edge cases it will not be.

Reply:
Regarding wrapping the `with` statement and returning the `Path`: once the `with` block exits the file at that path is no longer there if the file was temporarily created. Again, only a problem if the file is not on the filesystem.

Reply:
OK I think I'm getting the entire picture now. Being able to bundle a package and run as a zip was the part I was missing. In that case let's just leave this as-is for now. Thanks for your patience in explaining this!